cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -2281,52 +2454,102 @@ TEST(Cast, StructToDifferentSizeStruct) {
 
   EXPECT_RAISES_WITH_MESSAGE_THAT(
       TypeError,
-      ::testing::HasSubstr("Type error: struct field sizes do not match: struct<a: int8, "
-                           "b: int8> struct<a: int8, b: int8, c: int8>"),
+      ::testing::HasSubstr("struct fields don't match or are in the wrong order"),
       Cast(src, options));
 }
 
-TEST(Cast, StructToSameSizedButDifferentNullabilityStruct) {
-  // OK to go from not-nullable to nullable...
-  std::vector<std::shared_ptr<Field>> fields1 = {
-      std::make_shared<Field>("a", int8(), false),
-      std::make_shared<Field>("b", int8(), false)};
-  std::shared_ptr<Array> a1, b1;
-  a1 = ArrayFromJSON(int8(), "[1, 2]");
-  b1 = ArrayFromJSON(int8(), "[3, 4]");
-  ASSERT_OK_AND_ASSIGN(auto src1, StructArray::Make({a1, b1}, fields1));
-
-  std::vector<std::shared_ptr<Field>> fields2 = {
-      std::make_shared<Field>("a", int8(), true),
-      std::make_shared<Field>("b", int8(), true)};
-  std::shared_ptr<Array> a2, b2;
-  a2 = ArrayFromJSON(int8(), "[1, 2]");
-  b2 = ArrayFromJSON(int8(), "[3, 4]");
-  ASSERT_OK_AND_ASSIGN(auto dest1, StructArray::Make({a2, b2}, fields2));
-
-  CheckCast(src1, dest1);
-
-  // But not the other way around
-  std::vector<std::shared_ptr<Field>> fields3 = {
-      std::make_shared<Field>("a", int8(), true),
-      std::make_shared<Field>("b", int8(), true)};
-  std::shared_ptr<Array> a3, b3;
-  a3 = ArrayFromJSON(int8(), "[1, null]");
-  b3 = ArrayFromJSON(int8(), "[3, 4]");
-  ASSERT_OK_AND_ASSIGN(auto src2, StructArray::Make({a3, b3}, fields3));
-
-  std::vector<std::shared_ptr<Field>> fields4 = {
-      std::make_shared<Field>("a", int8(), false),
-      std::make_shared<Field>("b", int8(), false)};
-  const auto dest2 = arrow::struct_(fields4);
-  const auto options = CastOptions::Safe(dest2);
-
-  EXPECT_RAISES_WITH_MESSAGE_THAT(
-      TypeError,
-      ::testing::HasSubstr(
-          "Type error: cannot cast nullable struct to non-nullable "
-          "struct: struct<a: int8, b: int8> struct<a: int8 not null, b: int8 not null>"),
-      Cast(src2, options));
+TEST(Cast, StructToDifferentNullabilityStruct) {
+  {
+    // OK to go from non-nullable to nullable...
+    ARROW_SCOPED_TRACE("From non-nullable to nullable");
```

Comment:
nit, but this is really only useful for loops or templated tests, or when the line number is otherwise insufficient. For something like this it doesn't bring much.